glusterd2/pmap/registry.go

```diff
@@ -0,0 +1,207 @@
+package pmap
+
+import (
+	"expvar"
+	"fmt"
+	"net"
+	"sync"
+)
+
+const (
+	gfIanaPrivPortsStart = 49152
+	gfPortMax            = 65535
+)
+
+// PortState represents the type of state of the port
+type PortState int32
+
+// List of states an individual port can be in
+const (
+	GfPmapPortFree PortState = iota
+	GfPmapPortForeign
+	GfPmapPortLeased
+	GfPmapPortInUse
+)
+
+type portStatus struct {
+	Port  int       `json:"port"`
+	State PortState `json:"state"`
+	Conn  net.Conn  `json:"-"`
+}
+
+type brickStatus struct {
+	Port     int  `json:"port"`
+	SignedIn bool `json:"signed-in"`
+}
+
+type registryType struct {
+	sync.RWMutex
+	basePort  int
+	lastAlloc int
+	ports     [gfPortMax + 1]portStatus
+	bricks    map[string]*brickStatus
+
+	portLockFds map[int]int
+}
+
+func (r *registryType) init() {
+	for i := r.basePort; i <= gfPortMax; i++ {
+		// TODO: There is a race in this check when there are multiple
+		// glusterd2 instances running on the same machine.
+		if isPortFree(i) {
+			r.ports[i].State = GfPmapPortFree
+		} else {
+			r.ports[i].State = GfPmapPortForeign
+		}
+		r.ports[i].Port = i
+	}
+}
+
+// Allocate finds a free port and returns the port number. The
+// allocated free port will be in leased state until the brick
+// process it was leased to sends a SIGN IN request to glusterd2.
+func (r *registryType) Allocate(brickpath string) int {
+	r.Lock()
+	defer r.Unlock()
+
+	var port int
+	for p := r.basePort; p <= gfPortMax; p++ {
+		if r.ports[p].State == GfPmapPortFree || r.ports[p].State == GfPmapPortForeign {
```

Comment:
Shouldn't you be skipping PortForeign here?

Reply:
`PortForeign` was marked only during glusterd2's startup. That would've changed by a later time. In e2e test scenario with multiple gd2 instances, a `GfPmapPortForeign` could also mean that it's being used by another glusterd2 instance.